### PR TITLE
✨ Support /cosmos/tx/v1beta1/txs query endpoint

### PR DESCRIPTION
**Pull request by WeiJun0827**

### CODE REVIEW

File: rest/rest.go [hunk 1]
```diff
@@ -7,150 +7,323 @@ import (
...
+	if valueStr == "" {
+		return 0, nil
+	} else {
+		return strconv.ParseUint(valueStr, 10, 64)
+	}
 }
 
-func getAttributes(query url.Values) (attrs []Attribute) {
+func getPage(query url.Values) (uint64, error) {
+	page, err := getUint(query, "page")
```
Comment (nnkken) [inline rest/rest.go:38]: It seems better to separate the 2 APIs into separate files, so these helper functions won't mix together (which is confusing as I thought it should be `pagination.page` at the first glance).

So I suggest to have at least 3 files: one for legacy Amino endpoint, one for Stargate endpoint, one for common functionalities.

File: db/db.go [hunk 2]
```diff
@@ -168,43 +175,23 @@ func NewBatch(conn *pgxpool.Conn, limit int) Batch {
...
 			return err
 		}
 	}
-	txRes := TxResult{}
-	err := json.Unmarshal(txJSON, &txRes)
+	eventHashes := []int64{}
+	for _, log := range txRes.Logs {
+		eventHashes = append(eventHashes, util.GetEventHashes(log.Events)...)
+	}
+	txResJSON, err := encodingConfig.Marshaler.MarshalJSON(&txRes) // TODO fix "error": "unable to resolve type URL "
```
[the remark this answers is not in the document]
Comment (WeiJun0827) [reply db/db.go]: Just a temporary comment, removed.

File: rest/rest.go (hunk 1 above)
Comment (WeiJun0827) [reply rest/rest.go]: Got it, restructured, and ready for review.

File: db/db.go [hunk 3]
```diff
@@ -152,6 +161,72 @@ func GetLatestHeight(conn *pgxpool.Conn) (int64, error) {
...
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	res := make([]*types.TxResponse, 0, limit)
+	for rows.Next() {
+		var jsonb pgtype.JSONB
+		err := rows.Scan(&jsonb)
+		if err != nil {
+			return nil, err
```
Comment (nnkken) [inline db/db.go:218]: Should return more details on the error using `fmt.Errorf`, similar to line 223

File: poller/poller.go [hunk 4]
```diff
@@ -123,12 +114,17 @@ func poll(pool *pgxpool.Pool, ctx *CosmosCallContext, lastHeight int64) (int64,
...
+			url := fmt.Sprintf("%s/cosmos/tx/v1beta1/txs/%s", ctx.LcdEndpoint, txHash.String())
 			txResJSON, err := getResponse(ctx.Client, url)
 			if err != nil {
 				return 0, fmt.Errorf("cannot get tx response from lcd, error = %w, txhash = %s, height = %d, index = %d", err, txHash.String(), height, txIndex)
 			}
-			err = batch.InsertTx(txResJSON, height, txIndex)
+			txRes := txTypes.GetTxResponse{}
+			err = encodingConfig.Marshaler.UnmarshalJSON(txResJSON, &txRes)
+			if err != nil {
+				logger.L.Panicw("Cannot unmarshal tx response to JSON", "txhash", txHash, "tx_response", txResJSON, "error", err)
```
Comment (nnkken) [inline poller/poller.go:125]: Should not panic here but return error instead

File: rest/amino.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+		return
+	}
+	events, err := getEvents(q)
+	if err != nil {
+		c.AbortWithStatusJSON(400, gin.H{"error": err.Error()})
+		return
+	}
+	conn, err := db.AcquireFromPool(pool)
+	if err != nil {
+		logger.L.Panicw("Cannot acquire connection from database connection pool", "error", err)
```
Comment (nnkken) [inline rest/amino.go:71]: 🤔 I think here we should log with `Errorw` instead of `Panicw`, and then abort with 5xx

File: rest/router.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+	proxy := httputil.NewSingleHostReverseProxy(lcdURL)
+	proxyHandler := func(c *gin.Context) {
+		proxy.ServeHTTP(c.Writer, c.Request)
+	}
+
+	router := gin.New()
+	router.GET("/*endpoint", func(c *gin.Context) {
+		endpoint, ok := c.Params.Get("endpoint")
+		if !ok {
+			// ??? Gin bug?
```
Comment (nnkken) [inline rest/router.go:26]: Ah here we should add some log for error (so we can notice it and then report to Gin)

File: rest/router.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+		}
+		if endpoint == "/txs" {
+			handleAminoTxsSearch(c, pool)
+			return
+		}
+		if endpoint == "/cosmos/tx/v1beta1/txs" {
+			handleStargateTxsSearch(c, pool)
+			return
+		}
+		proxyHandler(c)
```
Comment (nnkken) [inline rest/router.go:38]: Prefer `switch` or `if ... else if ... else` in these multiple mutually exclusive cases, otherwise missing a `return` will cause unnoticed bug

File: rest/stargate.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/likecoin/likecoin-chain-tx-indexer/db"
+	"github.com/likecoin/likecoin-chain-tx-indexer/logger"
+)
+
+func getOffset(query url.Values) (uint64, error) {
+	return getUint(query, "pagination.offset")
+}
+
+func isOrderByDesc(query url.Values) (bool, error) {
```
Comment (nnkken) [inline rest/stargate.go:20]: Here you are hijacking a `bool` to represent an ordering, which I do not prefer since it is arbitrary and unfriendly to the reader (is `true` representing `ASC` or `DESC`?).

I prefer having something like an enum, e.g.
```Go
const (
    ORDER_ASC = "ASC"
    ORDER_DESC = "DESC"
)

func parseQueryOrder(query url.Values) (string, error) {
    // ...
    switch blablabla {
    case blablabla:
        return ORDER_ASC, nil
    case blablabla:
        return ORDER_DESC, nil
    default:
        return "", fmt.Errorf(blablabla)
    }
}
```

(In this case I used `string` as the underlying type, sometimes we may want even more complicated types, it depends on the coding context. I use `string` since eventually we want to plug it into SQL statement, so it would be convenient and easy to read)

File: rest/stargate.go [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+func trimSingleQuotes(s string) (string, error) {
+	if len(s) >= 2 {
+		if c := s[len(s)-1]; s[0] == c && c == '\'' {
+			return s[1 : len(s)-1], nil
+		}
+		return "", fmt.Errorf("expect query event value missing single quotes: %s", s)
+	}
+	return "", fmt.Errorf("invalid query event value: %s", s)
```
Comment (nnkken) [inline rest/stargate.go:39]: Prefer early return for error cases so the main code won't have to be indented:
```Go
if len(s) < 2 {
    return "", fmt.Errorf(blablabla)
}
// main code
```

File: rest/stargate.go [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+	case "ORDER_BY_DESC":
+		return true, nil
+	default:
+		return false, fmt.Errorf("available values for order_by: ORDER_BY_UNSPECIFIED, ORDER_BY_ASC, ORDER_BY_DESC")
+	}
+}
+
+func trimSingleQuotes(s string) (string, error) {
+	if len(s) >= 2 {
+		if c := s[len(s)-1]; s[0] == c && c == '\'' {
```
Comment (nnkken) [inline rest/stargate.go:34]: Hmm although the code is correct, I find this hard to follow, since it is saying `if the first character is the same as the last character and the last character is '\'', then...`, and I think `if the first character is '\'' and the last character is '\'', then...` is more straight forward

File: rest/stargate.go [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+		}
+		return "", fmt.Errorf("expect query event value missing single quotes: %s", s)
+	}
+	return "", fmt.Errorf("invalid query event value: %s", s)
+}
+
+func getEventMap(eventArray []string) (url.Values, error) {
+	m := make(url.Values)
+	for _, v := range eventArray {
+		if strings.Contains(v, "=") {
```
Comment (nnkken) [inline rest/stargate.go:45]: Same here, return errors early

File: rest/stargate.go [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+			arr := strings.SplitN(v, "=", 2)
+			value, err := trimSingleQuotes(arr[1])
+			if err != nil {
+				return nil, err
+			}
+			m[arr[0]] = []string{value}
```
Comment (nnkken) [inline rest/stargate.go:51]: What do you think we should do if the same event appears more than once?

File: rest/stargate.go [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+	if err != nil {
+		c.AbortWithStatusJSON(400, gin.H{"error": err.Error()})
+		return
+	}
+	limit, err := getLimit(q, "pagination.limit")
+	if err != nil {
+		c.AbortWithStatusJSON(400, gin.H{"error": err.Error()})
+		return
+	}
+	offsetInTimesOfLimit := offset / limit * limit // Cosmos' bug? 29 / 10 * 10 = 20
```
Comment (nnkken) [inline rest/stargate.go:72]: So the Cosmos API actually round the `offset` to a multiple of `limit`?

File: rest/stargate.go [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+	}
+	events, err := getEvents(eventMap)
+	if err != nil {
+		c.AbortWithStatusJSON(400, gin.H{"error": err.Error()})
+		return
+	}
+
+	conn, err := db.AcquireFromPool(pool)
+	if err != nil {
+		logger.L.Panicw("Cannot acquire connection from database connection pool", "error", err)
```
Comment (nnkken) [inline rest/stargate.go:92]: Same as above, maybe abort with 5xx

File: rest/stargate.go (hunk 13 above)
Comment (WeiJun0827) [reply rest/stargate.go:72]: Yes, for example, if `limit = 10`, there is no diff for `offset = 20` and `offset = 29`.

File: rest/stargate.go (hunk 10 above)
Comment (WeiJun0827) [reply rest/stargate.go:34]: I see, since it's a remaining function modify from determining none/single/double quotes, I did forget to use a more straightforward determination.

File: rest/stargate.go (hunk 8 above)
Comment (WeiJun0827) [reply rest/stargate.go:20]: I see, I thought I should prevent directly parse param to SQL statements `ASC` & `DESC` not until the Model layer (which is db.go) at the beginning. So I used a `bool`, but it's indeed arbitrary.

Comment (nnkken) [reply rest/stargate.go]: On the type level, you don't actually care about the underlying value (binary level / wrapped type / whatever), just like you don't care `'a'` is actually `97`, `0x61` or `0b01100001` inside the memory, all you know is that you can use `'a'` to refer to this character and do some operations (e.g. comparison, `toUpper` which outputs `'A'`).
In the case of indexer query, you only refer to those values in variable names and only need comparison (in switch case or if ... else if ... else), so the underlying values are not important, whether they are `true / false`, `"ASC" / "DESC"` or `1 / 2` or `1337 / 0xc0de`.
What is important in this level is that you have assigned them a name (in variables or constants, or enum if the language supports that), and the names are readable for others to understand. So you may actually use `true / false` but just wrap them into meaningful names and types.
You also would like to consider if you need extension for this enum-like type, e.g. if you use `bool` but then you have a 3rd option in the future, then you will need to modify all return types referring this enum-like type.
So usually we will use numbers as enum, since it is extensible, and is faster than string when doing comparison.
Using a string here is somewhat a hack, since in this case we want to eventually convert it into a string, so using string as underlying type will be a convenient choice.

**Review by nnkken — COMMENTED** [review]
Looks good to me, @williamchong any comments?

File: rest/router.go [hunk 15]
```diff
@@ -30,9 +30,9 @@ func Run(pool *pgxpool.Pool, listenAddr string, lcdEndpoint string) {
 		switch endpoint {
 		case "/txs":
 			handleAminoTxsSearch(c, pool)
-			return
 		case "/cosmos/tx/v1beta1/txs":
 			handleStargateTxsSearch(c, pool)
+		default:
 			return
```
Comment (nnkken) [inline rest/router.go]: The `proxyHandler(c)` must be in the `default` case, otherwise it won't be handled before return
